File: PO_py/program/Organisms/Animals/Antelope.py
import random
from program.Organisms.Animal import Animal

class Antelope(Animal):

    ANTELOPE_INITIATIVE = 4
    ANTELOPE_STRENGTH = 4

    def __init__(self, world, x, y):
        super().__init__(world, x, y, Antelope.ANTELOPE_INITIATIVE, Antelope.ANTELOPE_STRENGTH, 'A')
        print("Antelope created")

    def get_name(self):
        return "Antelope"

    def create_child(self, x, y):
        return Antelope(self.world, x, y)

    def randomize_position(self, move):
        rand = random.Random()
        move[0] = rand.randint(-1, 1)
        if move[0] == 0:
            move[1] = -1 if rand.random() < 0.5 else 1
        else:
            move[1] = rand.randint(-1, 1)
        if self.world.is_position_valid(self.x + (move[0] * 2), self.y + (move[1] * 2)):
            move[0] *= 2
            move[1] *= 2
            return True
        elif self.world.is_position_valid(self.x + move[0], self.y + move[1]):
            return True
        return False

    def dodge(self, attacker, defender):
        if defender.get_symbol() == 'A':
            rand = random.Random()
            if rand.random() < 0.5:  # 50% chance to dodge the attack
                print("Antelope avoids the attack")
                move = [0, 0]
                self.randomize_position(move)
                self.move(move[0], move[1])
                return True
        return False

File: PO_java/program/Organisms/Animals/Antelope.java
package projekt.program.Organisms.Animals;

import projekt.program.Organism;
import projekt.program.Organisms.Animal;
import projekt.program.World;

import java.util.Random;

public class Antelope extends Animal {

    private static final int ANTELOPE_INITIATIVE = 4;
    private static final int ANTELOPE_STRENGTH = 4;

    public Antelope(World world, int x, int y) {
        super(world, x, y, ANTELOPE_INITIATIVE, ANTELOPE_STRENGTH, 'A');
        System.out.println("Antelope created");
    }

    @Override
    public String getName() {
        return "Antelope";
    }

    @Override
    protected Organism createChild(int x, int y) {
        return new Antelope(this.world, x, y);
    }

    @Override
    public boolean randomizePosition(int[] move) {
        Random rand = new Random();
        move[0] = rand.nextInt(3) - 1;
        if (move[0] != 0) {
            move[1] = rand.nextInt(3) - 1;
        } else {
            move[1] = (rand.nextBoolean() ? -1 : 1);
        }
        if (world.isPositionValid(x + (move[0] * 2), y + (move[1] * 2))) {
            move[0] *= 2;
            move[1] *= 2;
            return true;
        } else if (world.isPositionValid(x + move[0], y + move[1])) {
            return true;
        }
        return false;
    }

    @Override
    public boolean dodge(Organism attacker, Organism defender) {
        if (defender.getSymbol() == 'A') {
            Random rand = new Random();
            if (rand.nextBoolean()) {
                System.out.println("Antelope avoids the attack");
                int[] move = new int[2];
                defender.randomizePosition(move);
                defender.move(move[0], move[1]);
                return true;
            }
        }
        return false;
    }
}

